File: 240628/마라톤 중간에 택시타기/taking-a-taxi-in-the-middle-of-the-marathon.py
N = int(input())
checkpoints = [list(map(int, input().split())) for _ in range(N)]

L = [0] * N 
R = [0] * N 

for i in range(1, N):
    a,b = checkpoints[i]
    c,d = checkpoints[i-1]
    L[i] = L[i-1] + abs(a-c) + abs(b-d)

for i in range(N-2, -1, -1):
    a,b = checkpoints[i+1]
    c,d = checkpoints[i]   
    R[i] = R[i+1] + abs(a-c) + abs(b-d)

ans = int(1e9)

for i in range(1, N-1):
    v = L[i-1] + R[i+1] +  abs(checkpoints[i+1][0] - checkpoints[i-1][0]) + abs(checkpoints[i+1][1] - checkpoints[i-1][1])
    ans = min(ans, v)
print(ans)

File: 240903/마라톤 중간에 택시타기/taking-a-taxi-in-the-middle-of-the-marathon.java
import java.util.*;

class Pair{
    int x, y;

    public Pair(int x, int y){
        this.x = x;
        this.y = y;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        Pair[] ps = new Pair[n];

        for (int i = 0; i < n; i++){
            int x = sc.nextInt();
            int y = sc.nextInt();

            Pair p = new Pair(x, y);
            ps[i] = p;
            
        }

        int[] L = new int[n];
        int[] R = new int[n];

        for (int i = 1; i < n; i++){
            int d = Math.abs(ps[i].x - ps[i-1].x) + Math.abs(ps[i].y - ps[i-1].y);
            L[i] = L[i-1] + d;
        }

        for (int i = n-2; i >= 0; i--){
            int d = Math.abs(ps[i].x - ps[i+1].x) + Math.abs(ps[i].y - ps[i+1].y);
            R[i] = R[i+1] + d;
        }

        // for (int i = 0; i < n; i++){
        //     System.out.print(L[i] + " ");
        // }
        // System.out.println();

        // for (int i = 0; i < n; i++){
        //     System.out.print(R[i] + " ");
        // }

        int ans = Integer.MAX_VALUE;

        for (int i = 1; i < n - 1; i++){
            int d = Math.abs(ps[i-1].x - ps[i+1].x) + Math.abs(ps[i-1].y - ps[i+1].y);

            ans = Math.min(ans, L[i-1] + R[i+1] + d);
        }
        System.out.println(ans);
    }
}